File: tests/go/secretmanager_test.go
package gcplocaltest

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"testing"

	"github.com/GuitarWag/gcp-local/tests/go/testutil"
)

func TestSecretManagerCRUDAndAccess(t *testing.T) {
	em := testutil.Start(t)
	base := "http://" + em.Host + "/v1/projects/" + project

	// create
	resp, body := doJSON(t, http.MethodPost, base+"/secrets?secretId=api-key", map[string]any{})
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("create: %d %s", resp.StatusCode, body)
	}

	// add version
	payload := base64.StdEncoding.EncodeToString([]byte("super-secret"))
	resp, body = doJSON(t, http.MethodPost, base+"/secrets/api-key:addVersion", map[string]any{
		"payload": map[string]any{"data": payload},
	})
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("addVersion: %d %s", resp.StatusCode, body)
	}

	// access latest
	resp, body = doJSON(t, http.MethodGet, base+"/secrets/api-key/versions/latest:access", nil)
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("access: %d %s", resp.StatusCode, body)
	}
	var ar struct {
		Payload struct {
			Data string `json:"data"`
		} `json:"payload"`
	}
	if err := json.Unmarshal(body, &ar); err != nil {
		t.Fatalf("decode: %v", err)
	}
	decoded, _ := base64.StdEncoding.DecodeString(ar.Payload.Data)
	if string(decoded) != "super-secret" {
		t.Errorf("payload = %q", string(decoded))
	}

	// list
	resp, body = doJSON(t, http.MethodGet, base+"/secrets", nil)
	if resp.StatusCode != http.StatusOK {
		t.Fatalf("list: %d %s", resp.StatusCode, body)
	}

	// delete
	resp, body = doJSON(t, http.MethodDelete, base+"/secrets/api-key", nil)
	if resp.StatusCode != http.StatusNoContent {
		t.Fatalf("delete: %d %s", resp.StatusCode, body)
	}
}

File: tests/typescript/secretmanager.test.ts
import { describe, expect, it } from "vitest";

const PROJECT = "local-project";

function host(): string {
  const h = process.env.GCP_LOCAL_HOST;
  if (!h) throw new Error("GCP_LOCAL_HOST not set");
  return h;
}

async function j(method: string, url: string, body?: unknown) {
  const res = await fetch(url, {
    method,
    headers: { "Content-Type": "application/json" },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  const text = await res.text();
  return { status: res.status, body: text ? JSON.parse(text) : undefined };
}

describe("secretmanager", () => {
  it("creates, adds version, accesses, deletes", async () => {
    const base = `http://${host()}/v1/projects/${PROJECT}/secrets`;
    let r = await j("POST", `${base}?secretId=ts-secret`, {});
    expect(r.status).toBe(200);

    const payload = Buffer.from("ts-classified").toString("base64");
    r = await j("POST", `${base}/ts-secret:addVersion`, { payload: { data: payload } });
    expect(r.status).toBe(200);

    r = await j("GET", `${base}/ts-secret/versions/latest:access`);
    expect(r.status).toBe(200);
    const decoded = Buffer.from(r.body.payload.data, "base64").toString();
    expect(decoded).toBe("ts-classified");

    r = await j("DELETE", `${base}/ts-secret`);
    expect(r.status).toBe(204);
  });

  it("missing secretId returns 400", async () => {
    const r = await j("POST", `http://${host()}/v1/projects/${PROJECT}/secrets`, {});
    expect(r.status).toBe(400);
  });
});
